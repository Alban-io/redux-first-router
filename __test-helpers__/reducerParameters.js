import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
