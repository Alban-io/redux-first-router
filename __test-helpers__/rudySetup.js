import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
