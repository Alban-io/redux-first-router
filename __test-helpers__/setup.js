import { applyMiddleware, createStore, compose } from 'redux'
import reduxThunk from 'redux-thunk'
import connectRoutes from '../src/connectRoutes'

const setup = (
  path = '/',
  options = { title: 'title', location: 'location' },
  routesMap
) => {
  routesMap = routesMap || {
    FIRST: '/first',
    SECOND: '/second/:param',
    THIRD: '/third'
  }

  options.initialEntries = path

  options.extra = { arg: 'extra-arg' }
  const tools = connectRoutes(routesMap, options)

  return { ...tools, routesMap }
}

export default setup

export const setupAll = (
  path,
  options,
  {
    rootReducer,
    preLoadedState,
    routesMap,
    dispatchFirstRoute = true,
    additionalMiddleware = store => next => action => next(action)
  } = {}
) => {
  const tools = setup(path, options, routesMap)
  const { middleware, reducer, enhancer, firstRoute } = tools
  const middlewares = applyMiddleware(
    reduxThunk,
    middleware,
    additionalMiddleware
  )
  const enhancers = compose(enhancer, middlewares)

  rootReducer =
    rootReducer ||
    ((state = {}, action = {}) => ({
      location: reducer(state.location, action),
      title: action.type
    }))

  const store = createStore(rootReducer, preLoadedState, enhancers)
  if (dispatchFirstRoute) store.dispatch(firstRoute())

  return {
    ...tools,
    store
  }
}
