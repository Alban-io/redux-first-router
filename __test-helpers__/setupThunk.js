import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
