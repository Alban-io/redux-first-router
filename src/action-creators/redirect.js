// @flow
import type { Action } from '../flow-types'

export default (action: Action, status: number = 302, committed: boolean = true) => {
  action.meta = action.meta || {}
  action.meta.location = action.meta.location || {}
  action.meta.location.kind = action.kind = 'redirect'
  action.meta.location.status = action.meta.location.status || status

  if (action.meta.location.committed === undefined) {
    action.meta.location.committed = committed // used to determine whether to do `history.redirect` or `history.push` (see utils/isRedirect.js)
  }

  return action
}
