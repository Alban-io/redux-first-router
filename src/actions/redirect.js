// @flow
import type { Action } from '../flow-types'

export default (action: Action, status: number = 302) => ({
  ...action,
  location: {
    ...action.location,
    status,
    url: null, // insures action can go through pipeline (`url` indicates its passed through already), see `utils/isTransformed.js`
    kind: 'redirect'
  }
})
