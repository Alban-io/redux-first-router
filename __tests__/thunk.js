import { setupAll } from '../__test-helpers__/setup'
import setupThunk from '../__test-helpers__/setupThunk'
import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'
import tempMock from '../__test-helpers__/tempMock'

import redirect from '../src/action-creators/redirect'

it('middleware:callThunk global', () => {
  const thunk = jest.fn(dispatch => {
    dispatch({ type: 'NON_ROUTE_ACTION' })
  })
  const { store } = setupThunk('/second/bar', thunk)
  expect(thunk).toBeCalled()
  expect(store.getState().location.type).toEqual('SECOND') // route stays the same (i.e. no redirect)
})

it('middleware:callThunk route', () => {
  const thunk = jest.fn()

  const routesMap = {
    FIRST: '/first',
    SECOND: {
      path: '/second',
      thunk
    },
    THIRD: '/third'
  }

  const { store } = setupAll('/first', undefined, { routesMap })
  store.dispatch({ type: 'SECOND' })

  expect(thunk).toBeCalled()
  expect(store.getState().location.type).toEqual('SECOND')
})

it('middleware:callThunk does NOT call thunk + change callback if isClientLoadSSR', () => {
  jest.resetModules()
  jest.doMock('../src/pure-utils/isClientLoadSSR', () => () => true)
  jest.doMock('../src/pure-utils/isServer', () => () => false)
  const { setupAll } = require('../__test-helpers__/setup')

  const thunk = jest.fn()
  const beforeEnter = jest.fn()
  const onComplete = jest.fn()

  const routeThunk = jest.fn()
  const routeBeforeEnter = jest.fn()
  const routeOnComplete = jest.fn()

  const routesMap = {
    FIRST: {
      path: '/first',
      beforeEnter: routeBeforeEnter,
      thunk: routeThunk,
      onComplete: routeOnComplete
    }
  }

  const options = {
    beforeEnter,
    thunk,
    onComplete
  }

  const setupOptions = { routesMap, dispatchFirstRoute: false }
  const { store, firstRoute } = setupAll('/first', options, setupOptions)

  store.dispatch(firstRoute())

  expect(beforeEnter).not.toBeCalled()
  expect(thunk).not.toBeCalled()
  expect(onComplete).not.toBeCalled()

  expect(routeBeforeEnter).not.toBeCalled()
  expect(routeThunk).not.toBeCalled()
  expect(routeOnComplete).not.toBeCalled()
})

it('middleware:callThunk DOES call thunk if locationState.kind !== "load"', () => {
  const thunk = jest.fn()
  const { store } = setupThunk('/first', thunk)

  const action = { type: 'SECOND', payload: { param: 'bar' } }
  store.dispatch(action)

  store.getState() /*? */

  expect(thunk).toBeCalled()
})

it('callThunk calls thunk with same `dispatch` argument as in middleware chain', () => {
  const thunk = jest.fn((dispatch, getState) => {
    const action = { type: 'THIRD', payload: { param: 'hurray' } }
    dispatch(action)

    return getState()
  })

  const { store } = setupThunk('/first', thunk)

  // thunk will be called
  const action = { type: 'SECOND', payload: { param: 'bar' } }
  store.dispatch(action)

  const { location } = store.getState() /*? $.location */

  // expect state matched that was dispatched in thunk
  expect(location.type).toEqual('THIRD')
  expect(location.pathname).toEqual('/third/hurray')
  expect(thunk).toHaveBeenCalled()
  expect(thunk.mock.calls[0].length).toEqual(3)
  expect(thunk.mock.calls[0][2].action).toMatchObject(action)
  expect(thunk.mock.calls[0][2].arg).toEqual('extra-arg')
})

it('pathless route calls callThunk', () => {
  const thunk = jest.fn()
  const routesMap = {
    FIRST: '/',
    PATHLESS: {
      thunk
    }
  }

  const { store, history } = setupAll('/', undefined, { routesMap })
  const action = { type: 'PATHLESS' }
  store.dispatch(action)

  expect(thunk).toHaveBeenCalled()
  expect(thunk.mock.calls[0].length).toEqual(3) // 2 args: dispatch, getState
  expect(thunk.mock.calls[0][2].action).toEqual(action)
})

it('pathless routes do not break other real route dispatches', () => {
  const thunk = jest.fn()
  const routesMap = {
    FIRST: '/',
    SECOND: '/second',
    PATHLESS: {
      thunk
    }
  }

  const { store, history } = setupAll('/', undefined, { routesMap })
  store.dispatch({ type: 'SECOND' })

  const { type } = store.getState().location
  expect(type).toEqual('SECOND')
})

it('pathless routes do not break history changes from real route dispatches', () => {
  const thunk = jest.fn()
  const routesMap = {
    FIRST: '/',
    SECOND: '/second',
    PATHLESS: {
      thunk
    }
  }

  const { store, history } = setupAll('/', undefined, { routesMap })
  history.push('/second')

  const { type } = store.getState().location
  expect(type).toEqual('SECOND')
})

it('CLIENT SPA: await dispatch(firstRoute())', async () => {
  jest.resetModules()
  jest.doMock('../src/pure-utils/isClientLoadSSR', () => () => false)
  jest.doMock('../src/pure-utils/isServer', () => () => false)

  const { setupAll } = require('../__test-helpers__/setup')

  const thunk1 = jest.fn(async dispatch => {
    await fakeAsyncWork()
    return dispatch({ type: 'THIRD' }) // key ingredient: users must insure their thunks await a redirect dispatch so its thunk (if also async) is awaited
  })

  const thunk2 = jest.fn(async dispatch => {
    await fakeAsyncWork()
    return 'thunk2called'
  })

  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second', thunk: thunk1 },
    THIRD: { path: '/third', thunk: thunk2 }
  }

  const setupOptions = { routesMap, dispatchFirstRoute: false }
  const { store, firstRoute } = setupAll('/second', undefined, setupOptions)

  // verify first route (and its thunk) has not been dispatched
  const { location } = store.getState()
  expect(location.pathname).toEqual('/second') // pathname will be correct because initialReducer state

  // but the thunk which redirects won't be called until now
  const res = await store.dispatch(firstRoute())
  const state = store.getState()

  // we'll only reach the state of the true first route here:
  expect(state.location.pathname).toEqual('/third')
  expect(state.location.type).toEqual('THIRD')
  expect(state.location.kind).toEqual('redirect')
  // expect(state.location).toMatchSnapshot()

  expect(thunk1.mock.calls.length).toEqual(1)
  expect(thunk2.mock.calls.length).toEqual(1)

  // verify second thunk is also called. This is dependent on the user returning their first thunk's
  // dispatch, which isnt required, unless they want the result in the following line like here:
  expect(res).toEqual('thunk2called')
})

it('CLIENT /w SSR: await dispatch(firstRoute()) -- no thunks etc called', async () => {
  jest.resetModules()
  jest.doMock('../src/pure-utils/isClientLoadSSR', () => () => true)
  jest.doMock('../src/pure-utils/isServer', () => () => false)

  const { setupAll } = require('../__test-helpers__/setup')

  const thunk = jest.fn()
  const beforeEnter = jest.fn()

  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second', thunk, beforeEnter }
  }

  const setupOptions = { routesMap, dispatchFirstRoute: false }
  const { store, firstRoute } = setupAll('/second', undefined, setupOptions)

  // verify first route (and its thunk) has not been dispatched
  const { location } = store.getState()
  expect(location.pathname).toEqual('/second') // pathname will be correct because initialReducer state
  expect(location.kind).toEqual('init')

  // but the thunk which redirects won't be called until now
  const action = await store.dispatch(firstRoute())
  const state = store.getState()

  expect(state.location.kind).toEqual('load')

  expect(beforeEnter.mock.calls.length).toEqual(0) // no beforeEnter called!
  expect(thunk.mock.calls.length).toEqual(0) // no thunks called!

  expect(action.kind).toEqual('load')
  expect(action.type).toEqual('SECOND')

  expect(action).toMatchSnapshot()
  expect(state.location).toMatchSnapshot()
})

it('SERVER: await dispatch(firstRoute()) -- redirected route not dispatched', async () => {
  jest.resetModules()
  jest.dontMock('../src/pure-utils/isClientLoadSSR')
  jest.doMock('../src/pure-utils/isServer', () => () => true)
  const { setupAll } = require('../__test-helpers__/setup')

  const thunk1 = jest.fn(async dispatch => {
    await fakeAsyncWork()
    return dispatch({ type: 'THIRD' })
  })

  const thunk2 = jest.fn()

  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second', thunk: thunk1 },
    THIRD: { path: '/third', thunk: thunk2 }
  }

  const setupOptions = { routesMap, dispatchFirstRoute: false }
  const { store, firstRoute } = setupAll('/second', undefined, setupOptions)

  // verify first route (and its thunk) has not been dispatched
  const { location } = store.getState()
  expect(location.pathname).toEqual('/second') // pathname will be correct because initialReducer state

  // but the thunk which redirects won't be called until now
  const action = await store.dispatch(firstRoute())
  const state = store.getState()

  // since it was a redirect on the server, the state will never change to that of the redirect
  expect(state.location.pathname).toEqual('/second')
  expect(state.location.type).toEqual('SECOND')
  expect(state.location.kind).toEqual('load')

  expect(thunk1.mock.calls.length).toEqual(1)
  expect(thunk2.mock.calls.length).toEqual(0) // second thunk never called!

  // instead, we get the action with enough info to short-circuit and call `res.redirect` on the server
  expect(action.kind).toEqual('redirect')
  expect(action.type).toEqual('THIRD')

  expect(action).toMatchSnapshot()
  expect(state.location).toMatchSnapshot()
})

it('dispatched thunk performs redirect with history.replace(path)', () => {
  const thunk = jest.fn(dispatch => {
    const action = redirect({
      type: 'THIRD',
      payload: { param: 'hurray' }
    }) /*? */
    dispatch(action)
  })
  const { store, history } = setupThunk('/first', thunk)

  const action = { type: 'SECOND', payload: { param: 'bar' } }
  store.dispatch(action)

  const { location } = store.getState() /*? */
  expect(location.kind).toEqual('redirect')
  expect(location.pathname).toEqual('/third/hurray')

  expect(history.length).toEqual(2) // if it wasn't a redirect, the length would be 3!
  expect(history.entries[1].pathname).toEqual('/third/hurray')
})
