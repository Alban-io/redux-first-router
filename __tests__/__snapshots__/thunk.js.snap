// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLIENT /w SSR: await dispatch(firstRoute()) -- no thunks etc called 1`] = `
Object {
  "kind": "load",
  "meta": Object {
    "location": Object {
      "current": Object {
        "kind": "load",
        "pathname": "/second",
        "payload": Object {},
        "type": "SECOND",
      },
      "history": undefined,
      "kind": "load",
      "prev": Object {
        "kind": "",
        "pathname": "",
        "payload": Object {},
        "type": "",
      },
    },
  },
  "payload": Object {},
  "type": "SECOND",
}
`;

exports[`CLIENT /w SSR: await dispatch(firstRoute()) -- no thunks etc called 2`] = `
Object {
  "hasSSR": undefined,
  "history": undefined,
  "kind": "load",
  "pathname": "/second",
  "payload": Object {},
  "prev": Object {
    "kind": "",
    "pathname": "",
    "payload": Object {},
    "type": "",
  },
  "routesMap": Object {
    "@@redux-first-router/NOT_FOUND": Object {},
    "FIRST": "/first",
    "SECOND": Object {
      "beforeEnter": [Function],
      "path": "/second",
      "thunk": [Function],
    },
  },
  "type": "SECOND",
}
`;

exports[`SERVER: await dispatch(firstRoute()) -- redirected route not dispatched 1`] = `
Object {
  "kind": "redirect",
  "meta": Object {
    "location": Object {
      "kind": "redirect",
    },
  },
  "type": "THIRD",
}
`;

exports[`SERVER: await dispatch(firstRoute()) -- redirected route not dispatched 2`] = `
Object {
  "hasSSR": true,
  "history": undefined,
  "kind": "load",
  "pathname": "/second",
  "payload": Object {},
  "prev": Object {
    "kind": "",
    "pathname": "",
    "payload": Object {},
    "type": "",
  },
  "routesMap": Object {
    "@@redux-first-router/NOT_FOUND": Object {},
    "FIRST": "/first",
    "SECOND": Object {
      "path": "/second",
      "thunk": [Function],
    },
    "THIRD": Object {
      "path": "/third",
      "thunk": [Function],
    },
  },
  "type": "SECOND",
}
`;
